FR #956: Provider-calculated split cost allocation (#1394)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/allocatedresourceid.md

# Allocated Resource ID

> An Allocated Resource ID is an identifier assigned by the provider which cost is being allocated to in a provider-calculated split cost allocation. The Allocated Resource ID is used to understand what the cost is being allocated to in [*charges*](#glossary:charge) where the provider is allocating costs to something other than the charge's [ResourceID](#ResourceId).

Reviewer: nit:

```suggestion
An Allocated Resource ID is an identifier assigned by the provider which cost is being allocated to in a provider-calculated split cost allocation. The Allocated Resource ID is used to understand what the cost is being allocated to in [*charges*](#glossary:charge) where the provider is allocating costs to something other than the *charge's* [ResourceID](#ResourceId).
```

> An Allocated Resource ID is an identifier assigned by the provider which cost is being allocated to in a provider-calculated split cost allocation. The Allocated Resource ID is used to understand what the cost is being allocated to in [*charges*](#glossary:charge) where the provider is allocating costs to something other than the *charge's* [ResourceID](#ResourceId).

Reviewer 2: Now that we have the supported feature in this PR as added by @mike-finopsorg, it would be cool to link to it.

HOWEVER.  The HTML generator doesn't do anything special for anchor links of supported features, so we can't do that for now.  I've spun up FR #1441 to do this in a future release.

So, you can resolve this comment -- just FYI!

Author: I'm aware he added it. I fixed some things on the supported feature last night and added it to the mdpp file so that the PDF would build.

Reviewer 2: I misunderstood how the anchors work when I wrote this.  We can currently use `#provider-calculatedsplitcostallocation`.  HOWEVER.  One other open thread on this PR proposes moving your `providercalculatedsplitcostallocation.md` in the columns folder and making it an official attribute, which I do believe you should do.  At that time, there would be a semantic conflict, and one of them will be suffixed with a `-1`.  (cc @mike-finopsorg)

In any case, I think we should just put this aside for now and not worry about it.  Resolving this thread.

The AllocatedResourceId column adheres to the following requirements:

* AllocatedResourceId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.
* AllocatedResourceId MUST be of type String.
* AllocatedResourceId MUST conform to [StringHandling](#stringhandling) requirements.

> * AllocatedResourceId nullability is defined as follows:
>   * AllocatedResourceId MUST be null when a charge is not related to a provider-calculated split cost allocation.
>   * AllocatedResourceId MUST NOT be null when a charge is related to a provider-calculated split cost allocation.

Reviewer: @kk09v @mike-finopsorg If I remember correctly, this is what we agreed on during the ad-hoc meeting on Sep 2. The wording of the requirements might not be optimal, but I hope the idea comes across.

```suggestion
* AllocatedResourceId nullability is defined as follows:
  * AllocatedResourceId MUST be null when a *charge* is not related to a provider-calculated split cost allocation.
  * AllocatedResourceId MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.
  * AllocatedResourceId MUST NOT be null when a *charge* represents the allocated portion of the origin *charge* assigned to a specific sub-resource.
```

Author: I did a lot of gymnastics to not refer to anything as a sub-resource and other than the column naming convention, tried to avoid the term resource when referring to {the entity or concept receiving the allocation}. There are many situations where these terms are not applicable (e.g. allocating cost to users).

I couldn't find a better way to say the suggestion on line 13, so I introduced it without the "assigned to a specific sub-resource." This can be reworded later or can stand as it is.

Reviewer: I was also weighing whether to include this or leave it out, and I was hoping you wouldn’t miss it - that is, that you’d make the call. Resolving the conversation.

* When AllocatedResourceId is not null, AllocatedResourceId adheres to the following additional requirements:

>   * AllocatedResourceId MUST be a locally unique identifier within the associated ResourceId at a point in time.

Reviewer 2: Per @mike-finopsorg: should this perhaps be "within a Charge Period"?

Author: Here was my thought process, using Kubernetes as an example:
- If the AllocatedResourceId is a pod, then it will be unique
- If the AllocatedResourceId is a namespace, then it is possible to remove a namespace and create a new namespace with the same name as the prior namespace within a charge period without violating the namespace uniqueness constraints of K8S.

If we use Charge Period, I'd like to relax it to SHOULD. To be honest, that would probably be fine because if there is a meaningful difference between them, that should be captured in AllocatedTags.

Reviewer 2: I agree with that, @kk09v.  So then, how about:
```suggestion
  * AllocatedResourceId SHOULD be a locally unique identifier within the associated ResourceId and ChargePeriod.
```

>   * AllocatedResourceId MAY NOT be unique across ResourceId or time periods.

Reviewer 2: Per @mike-finopsorg: same as previous comment and specify "Charge Period" more explicitly.

Reviewer 2: ```suggestion
  * AllocatedResourceId MAY NOT be unique across ResourceId or ChargePeriod values.
```

The AllocatedResourceId column adheres to the following requirements:

* AllocatedResourceId MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports provider-calculated split cost allocation.
* AllocatedResourceId MUST be of type String.
* AllocatedResourceId MUST conform to [StringHandling](#stringhandling) requirements.
* AllocatedResourceId nullability is defined as follows:
  * AllocatedResourceId MUST be null when a *charge* is not related to a provider-calculated split cost allocation.
  * AllocatedResourceId MUST be null when a *charge* represents the unallocated portion of the origin *charge* after split cost allocation.
  * AllocatedResourceId MUST NOT be null when a *charge* represents the allocated portion of the origin *charge*.
* When AllocatedResourceId is not null, AllocatedResourceId adheres to the following additional requirements:
  * AllocatedResourceId SHOULD be a locally unique identifier within the associated ResourceId and ChargePeriod.
  * AllocatedResourceId MAY NOT be unique across ResourceId or ChargePeriod values.

## Column ID

AllocatedResourceId

## Display Name

Allocated Resource ID

## Description

> Identifier assigned by the provider which cost will be allocated to in a provider-calculated split cost allocation.

Reviewer 2: ```suggestion
The identifier of the object to which cost is allocated in provider-calculated split cost allocation.
```
